mathesar/api/db/viewsets/queries.py

```diff
@@ -30,10 +31,22 @@ class QueryViewSet(
     pagination_class = DefaultLimitOffsetPagination
     filter_backends = (filters.DjangoFilterBackend,)
     filterset_class = UIQueryFilter
+    permission_classes = [IsAuthenticatedOrReadOnly]
     access_policy = QueryAccessPolicy
 
     def get_queryset(self):
-        queryset = self.access_policy.scope_queryset(self.request, UIQuery.objects.all())
+        queryset = None
+        if self.action == "results" or self.action == "retrieve" or self.action is None:
+            # Permission for both these actions are handled at QueryAccessPolicy
+            # Scoping is automatically handled by DRF since these are nested urls
+            # which contain the pk of the query
+            # It is essential to include None here, because it's called again by
+            # `filter_queryset` with action `None` after the query is formed
+            # TODO: Move these actions to separate viewsets to make this cleaner
+            queryset = UIQuery.objects.all().order_by('-created_at')
+        else:
+            queryset = self.access_policy.scope_queryset(self.request, UIQuery.objects.all())
```

Comment:
I'm having trouble understanding what's happening here. If permissions for these actions are handled in `QueryAccessPolicy`, why not use it here? Is scoping handled automatically, because the url contains the pk, which implies the necessary level of permission?

I trust that this code behaves correctly, but it's pretty hard to understand. You added a TODO comment for making this clearer. I think it would make sense to do that for this PR.

Reply:
@dmos62 

For both the actions that are listed here `results` and `retrieve`, we do not require scoping, since they request one particular query, which is identified since it's part of the url. `QueryAccessPolicy` takes care of checking if the user has access to that particular query.

> You added a TODO comment for making this clearer. I think it would make sense to do that for this PR.

The work required will need refactoring of this file and bunch of related files, which takes focus away from my current project. I'd like to take it up during the cooldown period or later. I'll create followup tickets for it. Does this sound good?

Reply:
Does the complexity stem from reusing the same viewset and the same access policy for both shared and regular queries? Are there disadvantages to using separate classes for both?

Unfortunately, I'm still having trouble reviewing this. For merging, in my opinion, there needs to be a refactor, or we can hop on a call and figure out how to make the code and documentation clearer without a refactor.

Reply:
The discussion was resolved over a call, and @dmos62 added a few commits to make this portion more readable.